tests/util/templated_test_case.hpp

```diff
@@ -0,0 +1,148 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2017 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+// Define a Catch test case templated on up to 20 types
+// The current type is exposed inside the test as `TestType`
+#define TEMPLATE_TEST_CASE(name, ...) \
```

Comment:
Just to make sure I'm following this correctly:

`TEMPLATE_TEST_CASE` expands into a function template with a unique name, whose implementation is the body of the block following the `TEMPLATE_TEST_CASE`, and a regular `TEST_CASE` with one `SECTION` per type, with each section containing a call to the function template with the type explicitly specified?